Add e2e tests to verify mixer report and check for ingress

bianpengyuan opened the pull request:
Fix #3418 

This test is duplicated from globalcheckandreport with ingress as destination in request count query.

istio-testing commented:
Hi @bianpengyuan. Thanks for your PR.

I'm waiting for a [istio](https://github.com/orgs/istio/people) member to verify that this patch is reasonable to test. If it is, they should reply with `/ok-to-test` on its own line. Until that is done, I will not automatically test new commits in this PR, but the usual testing commands by org members will still work. Regular contributors should join the org to skip this step.

I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).

<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>

bianpengyuan commented:
/assign @douglas-reid

douglas-reid commented:
/ok-to-test

douglas-reid commented on tests/e2e/tests/mixer/mixer_test.go:
```diff
@@ -305,6 +305,62 @@ func TestGlobalCheckAndReport(t *testing.T) {
+	}
+	allowPrometheusSync()
+
+	log.Info("Successfully sent request(s) to /productpage; checking metrics...")
```
instead of weaving together `t.Log` and `log.Info` statements, should we have consistency here of always using `t.Log`?

ldemailly commented:
/test e2e-suite-rbac-auth

istio-testing commented:
@bianpengyuan: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-suite-rbac-auth.sh | 71ec871002f87324b42d2e9ad3fa46de2f0b199d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3513/e2e-suite-rbac-auth/850/) | `/test e2e-suite-rbac-auth`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

ldemailly commented:
filled the unrelated failure issue #3600

mandarjog commented on tests/e2e/tests/mixer/mixer_test.go:
```diff
@@ -280,15 +280,71 @@ func TestGlobalCheckAndReport(t *testing.T) {
+	t.Logf("Baseline established: prior200s = %f", prior200s)
+	t.Log("Visiting product page...")
+
+	// visit production page as well as ingress.
```
nit: Visit product page through ingress.

mandarjog commented:
@bianpengyuan Thanks for the PR. 

1. This just adds report verification, we also need check.
2. Can you refactor code so that there is common code between `globalcheckandreport` and `IngressCheckAndReport`  instead of copying?

bianpengyuan commented:
@mandarjog Thanks for reviewing and sorry for the delay.

1. An ingress denial rule check test has been added for ingress. Let me know if this is enough or I need to add more for testing check.
2. For metric report, I extracted out the same code into a method. I am having a hard time to come up an elegant way to construct these tests. Let me know if you have a better idea. Also I have changed GlobalCheckAndReport to just GlobalReport -- there is no check rules at all in it.

mandarjog commented on tests/e2e/tests/mixer/mixer_test.go:
```diff
@@ -414,19 +425,25 @@ func TestNewMetrics(t *testing.T) {
 }
 
 func TestDenials(t *testing.T) {
+	testDenials(t, denialRule)
```
Please add 2 top level tests `func TestIngressDenial`  and `func TestDenial`.
This way there is better visibility about what exactly failed.

bianpengyuan replied:
Done.

mandarjog commented on tests/e2e/tests/mixer/mixer_test.go:
```diff
@@ -275,20 +285,21 @@ func TestGlobalCheckAndReport(t *testing.T) {
 	t.Logf("Baseline established: prior200s = %f", prior200s)
 	t.Log("Visiting product page...")
 
+	// visit production page.
```
nit: product page.    
I think you have used `production page` here and elsewhere.

bianpengyuan replied:
Done.

mandarjog commented on tests/e2e/tests/mixer/mixer_test.go:
```diff
@@ -250,16 +251,25 @@ func errorf(t *testing.T, format string, args ...interface{}) {
-	// establish baseline
-	t.Log("Establishing metrics baseline for test...")
-	query := fmt.Sprintf("istio_request_count{%s=\"%s\"}", destLabel, fqdn("productpage"))
+	checkMetricReport(t, promAPI, "productpage")
```
Similar to check Please add 2 top level tests for `TestMetric` and `TestIngressMetric`

bianpengyuan replied:
Done.

mandarjog commented:
Hey @bianpengyuan apart from my comments everything else looks good.
Using top level test names gives better visibility, thanks.

mandarjog commented:
/lgtm

mandarjog commented:
One last thing before merge, please update the PR title and description to reflect what it does now.

istio-testing commented:
@bianpengyuan: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-suite-rbac-auth.sh | 71ec871002f87324b42d2e9ad3fa46de2f0b199d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3513/e2e-suite-rbac-auth/850/) | `/test e2e-suite-rbac-auth`
prow/e2e-bookInfoTests.sh | 59713562b8bf0bd5bdb882672a5dbd13427d9917 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/3513/e2e-bookInfoTests/233/) | `/test e2e-bookInfo`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->
